hr_payroll_analysis/tests/test_payroll_analysis.py

```diff
@@ -0,0 +1,233 @@
+            cr, uid, self.payslip_ids, context=context)
+
+    def tearDown(self):
+        cr, uid, context = self.cr, self.uid, self.context
+
+        self.payslip_model.write(
+            cr, uid, self.payslip_ids, {'state': 'draft'}, context=context)
+
+        self.payslip_model.unlink(
+            cr, uid, self.payslip_ids, context=context)
+
+        self.contract_model.unlink(
+            cr, uid, [self.contract_id], context=context)
+
+        self.employee_model.unlink(
+            cr, uid, [self.employee_id], context=context)
+
+        self.structure_model.unlink(
+            cr, uid, [self.structure_id], context=context)
+
+        self.rule_model.unlink(
+            cr, uid, [self.rule_id, self.rule_2_id], context=context)
+
+        super(test_payroll_analysis, self).tearDown()
+
+    def test_payslip_analysis_line_generated(self):
+        cr, uid, context = self.cr, self.uid, self.context
+
+        self.payslip_model.process_sheet(
+            cr, uid, self.payslip_ids, context=context)
+
+        # Check that the analysis lines were created
+        line_ids = self.analysis_line_model.search(
+            cr, uid, [('payslip_id', 'in', self.payslip_ids)],
+            context=context)
+
+        lines = self.analysis_line_model.browse(
+            cr, uid, line_ids, context=context)
+
+        self.assertEqual(len(lines), 2)
+
+        for line in lines:
+            self.assertEqual(line.amount, 100)
+
+        # Process the payslips a second time
+        # Check that there is still one analysis line
+        # per payslip
+        self.payslip_model.process_sheet(
+            cr, uid, self.payslip_ids, context=context)
+
+        line_ids = self.analysis_line_model.search(
+            cr, uid, [('payslip_id', 'in', self.payslip_ids)],
+            context=context)
+
+        self.assertEqual(len(line_ids), 2)
+
+        # Cancel a payslip and check that there is only one
+        # analysis line left
+        self.payslip_model.cancel_sheet(
+            cr, uid, [self.payslip_ids[0]], context=context)
+
+        line_ids = self.analysis_line_model.search(
+            cr, uid, [('payslip_id', 'in', self.payslip_ids)],
+            context=context)
+
+        lines = self.analysis_line_model.browse(
+            cr, uid, line_ids, context=context)
+
+        self.assertEqual(len(lines), 1)
+        self.assertEqual(lines[0].payslip_id.id, self.payslip_ids[1])
+
+    def test_salary_rule_refresh(self):
+        cr, uid, context = self.cr, self.uid, self.context
+
+        self.payslip_model.process_sheet(
+            cr, uid, self.payslip_ids, context=context)
+
+        # Check that the analysis lines were created
+        line_ids = self.analysis_line_model.search(
+            cr, uid, [('payslip_id', 'in', self.payslip_ids)],
+            context=context)
+
+        lines = self.analysis_line_model.browse(
+            cr, uid, line_ids, context=context)
+
+        self.assertEqual(len(lines), 2)
```

Comment:
The lines from 196 to 207 are equals to the lines from 150 to 161... you can extract these lines in a method called by the two tests.